daemon/fmutex/fmutex.go

```diff
@@ -0,0 +1,65 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package fmutex implements locking for the snappy daemon
+package fmutex
+
+import (
+	"sync"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/logger"
+	"github.com/ubuntu-core/snappy/priv"
+)
+
+// FMutex gives you an exclusive mutex that also holds the filesystem lock.
+type FMutex struct {
+	mutex *sync.Mutex
+	flock *priv.FileLock
+}
+
+var _ sync.Locker = (*FMutex)(nil)
+
+// New returns a brand new FMutex, ready for use.
+func New() *FMutex {
+	return &FMutex{
+		mutex: &sync.Mutex{},
+		flock: priv.NewFileLock(dirs.SnapLockFile),
```

Comment:
If we take path in (as in New(path string)) this implementation becomes unbound to any particular file, and it also stops trusting on any global state. It's more natural for daemon then to know where its own paths lie, so we can ask it to input into this function. I hope eventually we can move in the dirs package closer to the daemon and stop trusting on it as global state.